draft-ietf-tls-external-psk-importer.md

```diff
@@ -252,14 +264,21 @@ This does not affect the KDF operation used to derive Imported PSKs.
 # Incremental Deployment {#rollout}
 
```

Comment:
Hrm. I'm very confused what this section is saying, with both new and old text. :-) Are we saying that you take the EPSK and plug it directly into your TLS 1.2 API, without any derivation step at all? I guess that's the only way existing applications can avoid rekeying... but I'm not sure the safety claims hold, at least not obviously.

We'd need to confirm that the TLS 1.2 PRF is separated from our derivation construction. Also TLS 1.2 doesn't not lock things to algorithms, so we'd also have to believe that SHA-256, SHA-384, and MD5/SHA-1 are sufficiently unrelated to the EPSK KDF hash. (Which, let's be honest, is probably true. But if we were okay with that, we wouldn't have gotten ourselves into this mess. :-) )

Or am I completely misunderstanding what this is saying?

Edit: Oh, I guess we have some text on this in security considerations. Though this section reads much more certain than security considerations does.

Reply:
Yeah, you basically have it right. (Matches my mental model.) Feel free to suggest alternative phrasings if you have them, or maybe we just drop a pointer to the security considerations? I don't feel strongly either way.

Reply:
I guess I would have just dropped this, though I also don't care very much about existing TLS 1.2 uses and others probably do.

BTW, this section also contradicts this text:
>    Endpoints which import external keys MUST NOT use either the external
>    keys or the derived keys for any other purpose

On alternate phrasings, I guess the question is what we want to say. We could say any of:

1. You're not allowed to use EPSKs directly in TLS 1.2. It's not safe. If you have an existing TLS 1.2 PSKs, you've got to rekey, too bad.
2. You're allowed to derive IPSKs from EPSKs for use with TLS 1.2. This doesn't help existing TLS 1.2 PSKs, and new things should use TLS 1.3, so this seems not useful.
3. You're allowed to take a TLS 1.2 PSK and use it as an EPSK directly. This contradicts the MUST NOT above and is not obviously safe. So maybe we say it's NOT RECOMMENDED and tell people to quickly migrate TLS 1.2 off? Though that's kinda RFC6919-y.
4. Same as above but we claim it's safe. We'd probably need someone to do some analysis here.
5. Define a way to derive an EPSK from a TLS 1.2 PSK using the TLS 1.2 PRF, in hopes that it is more clearly separated. Though even that is slightly a lost cause because TLS 1.2 PSKs were themselves not clearly separated. This gives a transition path, but it means that, even when your system removes TLS 1.3, you've still got that extra derivation step in there.

Incidentally, my [old draft](https://datatracker.ietf.org/doc/html/draft-davidben-tls-universal-psk-00#section-3) did (5).

Which do you think we should do?

Reply:
I think we should aim for minimal changes to TLS 1.2 at this point, so I'd be inclined to go with (3) or (4).

Reply:
Well I think only (2) involves changes to TLS 1.2. It's also not really useful, so we can discard it.

(1) simply says that your `ConfigureTLS12PSK` API doesn't do anything useful with PSK importers, and that you MUST (BUT WE KNOW YOU WON'T) use separate PSKs with TLS 1.3. For instance, your PSK issuance logic could issue parallel keys for TLS 1.2 and EPSKs. But this means you cannot enable TLS 1.3 transparently on clients that haven't gotten parallel PSKs issued. Depending on your application, this may be easy or very inconvenient.

(3) and (4) say that your `ConfigureTLS12PSK` API (optionally? by default?) also is a `ConfigureEPSK` API and we'll reuse the keys. This means you can enable TLS 1.3 transparently, but it's not really sound and you OUGHT TO get rid of TLS 1.2 in your application.

(5) says that, if you call `ConfigureEPSK`, you get the EPSK thing in this draft. If you call `ConfigureTLS12PSK(psk)`, as a compatibility scheme, the library will internally also call `ConfigureEPSK(DeriveEPSKFromTLS12PSK(psk))`. This means you can enable TLS 1.3 transparently, but when you drop TLS 1.2 support, you'll still need to go through the `DeriveEPSKFromTLS12PSK` codepath until you, like (1), drive fresh keys.

Reply:
Oh, I guess I misunderstood you. If (5) is just _using_ the TLS 1.2 PRF, that would be fine, I guess. This will require more invasive changes to the draft for 1.2, though, which is... a bummer. Do you think you'd have cycles to do that? (I'm not sure when I'd be able to get to it.)

Reply:
Yeah, I'm not sure whether (5) is worth it. I guess it's also a pretty big change to be doing this late. I don't feel super strongly about what to do here... my immediate comment is just that the current text can't seem to decide between (3) and (4). And that, if we assert it's safe, it should actually be safe. :-)

Cycles-wise, not sure when I'd be able to get to it either. (Right now I really want to finish that ECH implementation.)

Reply:
Thanks @kaduk! I refactored this suggestion and applied it. I'll aim to submit tomorrow before the deadline. Note that we may want to pause this document to collect implementation experience and additional analysis.